Sources/BuildServerIntegration/BuildServerManager.swift

```diff
@@ -962,6 +962,96 @@ package actor BuildServerManager: QueueBasedMessageHandler {
+          ).uri
+          textEdit.textDocument.uri = newUri
+          newDocumentChanges.append(.textDocumentEdit(textEdit))
+        case .createFile(var create):
+          let newUri = self.locationAdjustedForCopiedFiles(
+            Location(uri: create.uri, range: Position(line: 0, utf16index: 0)..<Position(line: 0, utf16index: 0))
+          ).uri
+          create.uri = newUri
+          newDocumentChanges.append(.createFile(create))
+        case .renameFile(var rename):
+          let newOldUri = self.locationAdjustedForCopiedFiles(
+            Location(uri: rename.oldUri, range: Position(line: 0, utf16index: 0)..<Position(line: 0, utf16index: 0))
+          ).uri
+          let newNewUri = self.locationAdjustedForCopiedFiles(
+            Location(uri: rename.newUri, range: Position(line: 0, utf16index: 0)..<Position(line: 0, utf16index: 0))
+          ).uri
+          rename.oldUri = newOldUri
+          rename.newUri = newNewUri
+          newDocumentChanges.append(.renameFile(rename))
+        case .deleteFile(var delete):
+          let newUri = self.locationAdjustedForCopiedFiles(
+            Location(uri: delete.uri, range: Position(line: 0, utf16index: 0)..<Position(line: 0, utf16index: 0))
+          ).uri
+          delete.uri = newUri
+          newDocumentChanges.append(.deleteFile(delete))
+        }
+      }
+      edit.documentChanges = newDocumentChanges
+    }
+    return edit
+  }
+
+  package func locationsOrLocationLinksAdjustedForCopiedFiles(_ response: LocationsOrLocationLinksResponse?) async -> LocationsOrLocationLinksResponse? {
+    guard let response = response else {
+      return nil
+    }
+    switch response {
+    case .locations(let locations):
+      let remappedLocations = await self.locationsAdjustedForCopiedFiles(locations)
+      return .locations(remappedLocations)
+    case .locationLinks(let locationLinks):
+      var remappedLinks: [LocationLink] = []
+      for link in locationLinks {
+        let newUri = self.locationAdjustedForCopiedFiles(Location(uri: link.targetUri, range: link.targetRange)).uri
+        remappedLinks.append(LocationLink(
+          originSelectionRange: link.originSelectionRange,
+          targetUri: newUri,
+          targetRange: link.targetRange,
+          targetSelectionRange: link.targetSelectionRange
+        ))
+      }
+      return .locationLinks(remappedLinks)
+    }
+  }
+
+  package func typeHierarchyItemAdjustedForCopiedFiles(_ item: TypeHierarchyItem) async -> TypeHierarchyItem {
+    let adjustedLocation = await self.locationAdjustedForCopiedFiles(Location(uri: item.uri, range: item.range))
+    return TypeHierarchyItem(
+      name: item.name,
+      kind: item.kind,
+      tags: item.tags,
+      detail: item.detail,
+      uri: adjustedLocation.uri,
+      range: adjustedLocation.range,
+      selectionRange: item.selectionRange,
```

Comment:
It might seem redundant now but I’d like to run the selection range through `locationAdjustedForCopiedFiles` as well, just so we have a centralized place that might also adjust the range eg. if it discovers that copying the file removed the first line and thus all line number need to be shifted by 1 or something like that (not proposing that we do this now but it might be something we could consider doing in the future).